### PR TITLE
Support training with mixed precision (`float16`, `bfloat16`)

### PR DESCRIPTION
**Pull request by nucli-vicky**
See #1205 

Also fixes bug wrt model being used in `.train()` mode during eval when `compute_eval_losses=True`

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` and `ruff check .` run successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

**Comment by nucli-vicky** [comment]
### To show this implementation trains properly, and gives comparable runs between `use_amp`  `True` and `False`. 

Small test training run for 10 epochs on DIV2K 2x SR with `SRResNet`:

Test with `ckp_9.pth.tar` on DIV2K val: 

Trained without AMP:

- PSNR: 27.514 +- 3.858
- SSIM: 0.818 +- 0.106


Trained with AMP (`use_amp=True`):

- PSNR: 28.139 +- 4.200
- SSIM: 0.819 +- 0.104

**Comment by nucli-vicky** [comment]
**Training curve when not using AMP:**

image

**Training curve when using AMP:**

image

**Comment by nucli-vicky** [comment]
While writing this, I came across the following:

`compute_loss` is called both when doing eval, and when doing inference with `compute_eval_losses`.

However, in there, it will *always* call `self.model_inference(` with `train=True`. 

This causes 

(i) no `torch.no_grad` to speed up eval.

(ii) when `compute_eval_losses`, both the eval losses and metrics will use an `x_net` obtained under `net.train()`: with dropout active and `batchnorm` layers not in eval mode either.

File: deepinv/training/trainer.py [hunk 1]
```diff
@@ -68,6 +69,7 @@ class Trainer:
     :param bool online_measurements: Generate new measurements `y` in an online manner at each iteration by calling
         `y=physics(x)`. If `False` (default), the measurements are loaded from the training dataset.
     :param str, torch.device device: Device on which to run the training (e.g., 'cuda', 'mps' or 'cpu'). Default is first 'cuda' and second 'mps' if available, otherwise 'cpu'.
+    :param bool use_amp: Whether to use automatic mixed precision (float16). Currently, will only be used during training steps, not eval or test.
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py]: I would use a general parameter here, for example `mixed_precision: str | None = "no"` with choice between "no" (or "fp32"), "fp16" or "bf16". "bf16" is interesting since it has the same dynamic range as fp32 so no need for GradScaler

**Review by mh-nguyen712 — CHANGES_REQUESTED** [review]
Thanks for the initiative! I think it's more flexible the leave user chose the compute dtype between fp32, fp16 or bfp16 rather than hardcoded it

File: deepinv/training/trainer.py [hunk 2]
```diff
@@ -862,19 +878,30 @@ def compute_loss(self, physics, x, y, train=True, epoch: int = None, step=False)
 
         if train or self.compute_eval_losses:
             # Evaluate reconstruction network
-            x_net = self.model_inference(y=y, physics=physics, x=x, train=True)
+            x_net = self.model_inference(y=y, physics=physics, x=x, train=train)
```
Comment (nucli-vicky) [inline deepinv/training/trainer.py]: Turns out this is what's breaking the tests. I will revert it, and let this be handled by a future PR, as this one has been quite stale for a while.

**Comment by nucli-vicky** [comment]
@mh-nguyen712 thanks for the suggestion! I incorporated it here now :) [finally]

**Comment by Andrewwango** [comment]
re compute_eval_losses, this is reported in #1096, but I don't think there is a trivial fix because some losses can only be calculated in train mode eg ReducedResolutionLoss

**Comment by Andrewwango** [comment]
Hey @nucli-vicky , thanks for the contribution, and except for fp16, I think it's neat and concise addition.

Re. fp16, I think the fp16 support with gradscaler is quite risky as it potentially introduces lots of uncaught silent PyTorch bugs to a new user (e.g. https://github.com/pytorch/pytorch/issues/58228 ) - it would be unfortunate for such a user to see our neatly wrapped fp16 support and trust it 100%. I think for fp16, if we support it at all, it should come with clear warnings that it can lead to hidden bugs, and is included for legacy, and users should really use bf16 instead.

Secondly, it would be good to have a similar plot to the above using bf16 where the training should be pretty much identical.

**Comment by nucli-vicky** [comment]
Thanks @Andrewwango 

Re `compute_eval_losses`: Seems I missed that issue, thanks! But I think it's about more than efficiency, for nets with batchnorm, the eval results (and also metrics)  will be way different when `compute_eval_losses` is on, as the batch statistics will be used for validation instead of the population statistics tracked during training. Even mathematical properties can be different (e.g. for certain convnets with batchnorm, the scale-equivariance property may hold in val mode, but not in train mode)

Re float16: I think the issues relate only/mostly to the training of complex dtype networks / data. Maybe we can narrow the warning down to that?

**Comment by Andrewwango** [comment]
@nucli-vicky re. compute_eval_losses: it makes sense and I agree with you. Just need to find a way to deal with losses that can only be evaluated in train mode... perhaps we need to fix those losses directly, more specifically their adapted model wrappers (e.g. the model wrapper forward pass branches train vs eval mode as before, but in the loss calculation, there is no branch and only train mode behaviour is used)

re. float16: I don't know much about the various pitfalls (outside complex data). Maybe to mitigate the risk, it would be good to have some tests or examples on various common deepinv training scenarios to check that there are no performance shocks.

**Comment by nucli-vicky** [comment]
@Andrewwango Thanks for your thoughts! Finally got to finalizing this PR 😅 

Re. `compute_eval_losses`: agree with you here. I'm not sure if we should wait for the full Trainer refactor to address this, or might be a good thing to take care of during the Hackathon?

Re. float16: From my experience, I've never encountered issues due to `float16` AMP across a variety of tasks and model types. While I personally don't expect any issues (aside from complex dtype models), we can't really be sure across the wide range of models, tasks, and data that `deepinv`could get used for. Therefore I've added a warning that notes AMP as an experimental feature, and encouraging users to report any issues/bugs they may encounter.

File: deepinv/training/trainer.py [hunk 3]
```diff
@@ -69,6 +71,8 @@ class Trainer:
     :param bool online_measurements: Generate new measurements `y` in an online manner at each iteration by calling
         `y=physics(x)`. If `False` (default), the measurements are loaded from the training dataset.
     :param str, torch.device device: Device on which to run the training (e.g., 'cuda', 'mps' or 'cpu'). Default is first 'cuda' and second 'mps' if available, otherwise 'cpu'.
+    :param bool, str mixed_precision: Mixed precision to use. If False, standard float32 training is performed. If True, defaults to float16.
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py:75]: I would allow any string in this list: `float32`, `fp32`, `float16`, `fp16`, `bfloat16`, `bf16`

Comment (nucli-vicky) [reply deepinv/training/trainer.py:75]: thanks, good point to also include the `bf16` and `fp16` abbreviations :)

Im not sure as to including float32: in that case, there is no mixed precision, its just regular, single-precision training. So I think being able to specify float32 it for `mixed_precision` is wrong semantically. What do u think?

File: deepinv/training/trainer.py [hunk 4]
```diff
@@ -653,6 +683,11 @@ def check_clip_grad(self):
 
         return grad_norm
 
+    def get_autocast_context(self):
+        if self.mixed_precision in [torch.float16, torch.bfloat16]:
+            return autocast(device_type=self.device.type, dtype=self.mixed_precision)
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py:715]: Make sure that `self.device` is a `torch.device`, since user might pass a string as device. I think it's safer to make sure this in `__post_init__`

File: deepinv/training/trainer.py [hunk 5]
```diff
@@ -872,14 +922,22 @@ def compute_loss(self, physics, x, y, train=True, epoch: int = None, step=False)
...
-            norm = self.check_clip_grad()
-            if norm is not None:
-                logs["gradient_norm"] = self.check_grad_val.avg
+            # GradScaler forbids calling unscale_() more than once per optimizer between step() calls.
+            defer_clip = (
+                self.scaler.is_enabled()
+                and self.optimizer_step_multi_dataset
+                and not step
+            )
+            if not defer_clip:
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py:960]: I think there is an inconsistent gradient clipping behavior between AMP and non-AMP if `optimizer_step_multi_dataset=True`:
- If AMP: clipping is deferred until gradients across all datasets accumulate (clip(g1 + g2)).
- No AMP: defer_clip is False, so check_clip_grad() runs for each dataset individually before accumulation (clip(g1) + clip(g2)).
  
I think we should be clear on that

File: deepinv/training/trainer.py [hunk 6]
```diff
@@ -1335,7 +1399,17 @@ def train(
         :returns: The trained model.
         """
         self.setup_train()
-
+        if self.mixed_precision and self.device.type != "cuda":
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py]: I think it's cleaner to move these warnings to `__post_init__`

**Review by mh-nguyen712 — CHANGES_REQUESTED** [review]
Thanks for the updates! Here are my last comments and hope that we can merge this before the hackathon!

File: deepinv/training/trainer.py [hunk 7]
```diff
@@ -559,6 +585,8 @@ def setup_train(self, train: bool = True, **kwargs):
         except (ValueError, TypeError, AttributeError):
             self._model_accepts_update_parameters = False
 
+        self.scaler = GradScaler(enabled=self.mixed_precision == torch.float16)
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py]: I think it's better to precise the `device` in the GradScaler. 
```suggestion
        self.scaler = GradScaler(
            device=self.device.type, 
            enabled=self.mixed_precision == torch.float16 
        )
```

File: deepinv/training/trainer.py [hunk 8]
```diff
@@ -1335,7 +1399,17 @@ def train(
         :returns: The trained model.
         """
         self.setup_train()
-
+        if self.mixed_precision and self.device.type != "cuda":
+            warnings.warn(
```
Comment (mh-nguyen712) [inline deepinv/training/trainer.py]: I think `bfloat16` is natively supported by modern CPU too, eg: https://intel.github.io/intel-extension-for-pytorch/cpu/latest/tutorials/features/amp.html

File: deepinv/training/trainer.py (hunk 3 above)
Comment (mh-nguyen712) [reply deepinv/training/trainer.py:75]: Yes agree!

Comment (nucli-vicky) [reply deepinv/training/trainer.py:75]: added fp16 and bf16

File: deepinv/training/trainer.py (hunk 5 above)
Comment (nucli-vicky) [reply deepinv/training/trainer.py:960]: fair! Unfortunately there is right now not really a way to do `clip(g1) + clip(g2)` for `float16` as it would require calling unscale twice which is supported. 

Wondering where best to document this discrepancy?

File: deepinv/training/trainer.py (hunk 4 above)
Comment (nucli-vicky) [reply deepinv/training/trainer.py:715]: I think prior to this PR, code passing as device a string would not break. So are u proposing we convert `device` variable to a `torch.device` in `__post_init__`? 

Will do

File: deepinv/training/trainer.py (hunk 7 above)
Comment (nucli-vicky) [reply deepinv/training/trainer.py]: thanks!

File: deepinv/training/trainer.py (hunk 8 above)
Comment (nucli-vicky) [reply deepinv/training/trainer.py]: true, but this warning also mainly serves for other accelerators (MPS for instance). 

As mixed precision is currently only implemented for training (which users will typically not do on CPU) and a lot of CPUs do not have native `bfloat16`, I think it's okay to keep the warning as is?

**Comment by nucli-vicky** [comment]
Thanks @mh-nguyen712 for the comments, tried to address most of them.

File: deepinv/training/trainer.py (hunk 5 above)
Comment (mh-nguyen712) [reply deepinv/training/trainer.py:960]: I think we can add a note to the main docstring of the Trainer class. Can you add this? Then I think the PR is ready to be merged! Thanks for you works!

Comment (nucli-vicky) [reply deepinv/training/trainer.py:960]: note added!

**Review by mh-nguyen712 — APPROVED** [review]
